File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    unordered_map<int, int> inorderMap;

    TreeNode* build(vector<int>& inorder,
                    vector<int>& postorder,
                    int inStart,
                    int inEnd,
                    int postStart,
                    int postEnd) {

        if (inStart > inEnd || postStart > postEnd)
            return NULL;

        int rootValue = postorder[postEnd];

        TreeNode* root = new TreeNode(rootValue);

        int rootIndex = inorderMap[rootValue];

        int leftSize = rootIndex - inStart;

        root->left = build(inorder,
                           postorder,
                           inStart,
                           rootIndex - 1,
                           postStart,
                           postStart + leftSize - 1);

        root->right = build(inorder,
                            postorder,
                            rootIndex + 1,
                            inEnd,
                            postStart + leftSize,
                            postEnd - 1);

        return root;
    }

    TreeNode* buildTree(vector<int>& inorder,
                        vector<int>& postorder) {

        int n = inorder.size();

        for (int i = 0; i < n; i++) {

            inorderMap[inorder[i]] = i;
        }

        return build(inorder,
                     postorder,
                     0,
                     n - 1,
                     0,
                     n - 1);
    }
};

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private int postIdx;
    private HashMap<Integer, Integer> map;

    public TreeNode buildTree(int[] inorder, int[] postorder) {

        map = new HashMap<>();

        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }

        postIdx = postorder.length - 1;

        return build(inorder, postorder, 0, inorder.length - 1);
    }

    private TreeNode build(int[] inorder, int[] postorder,
                           int left, int right) {

        if (left > right) {
            return null;
        }

        int rootVal = postorder[postIdx--];
        TreeNode root = new TreeNode(rootVal);

        int idx = map.get(rootVal);

        // Build right subtree first
        root.right = build(inorder, postorder, idx + 1, right);

        // Then left subtree
        root.left = build(inorder, postorder, left, idx - 1);

        return root;
    }
}